Core/Phases: Fix SetInPhase not honoring zone phase

roc13x opened the pull request:
**Changes proposed:**
- Fix SetInPhase not honoring zone phases, only area
- Update area and zone phases on quest update

**Target branch(es):**
- [x] master

**Tests performed:**
- Compiled
- Tested in-game

roc13x committed:
Core/Phases: Fix SetInPhase not honoring zone phase

Also update area phases on quest update

Shauren commented on src/server/game/Entities/Object/Object.cpp:
```diff
@@ -3028,11 +3028,13 @@ bool WorldObject::SetInPhase(uint32 id, bool update, bool apply)
-                        if (sConditionMgr->IsObjectMeetToConditions(this, phase.Conditions))
-                            return false;
+            PhaseInfo const& phases = sObjectMgr->GetAreaAndZonePhases();
+            for (PhaseInfo::const_iterator itr = phases.begin(); itr != phases.end(); ++itr)
```
I don't like this - couldn't you make another store in ObjectMgr to lookup phase by id?

roc13x replied:
Wouldn't that require quite a bit of nearly-duplicate code to construct a similar store? Especially in ConditionMgr::addToPhases

Shauren replied:
No, I mean something like `map<phaseId, vector<PhaseInfoStruct const*>>` in ObjectMgr (constructed in same place as `_phases`)

Shauren commented on src/server/game/Entities/Player/Player.cpp:
```diff
@@ -16005,7 +16005,7 @@ void Player::SendQuestUpdate(uint32 questId)
 
     UpdateForQuestWorldObjects();
-    SendUpdatePhasing();
+    UpdateAreaAndZonePhase();
```
This was previously a target of discussion in another PR - by forcing the update here you break quests that should not update phase here

roc13x committed:
Core/Phase: Refactor area phase storage

Map by phaseid instead of area. No function actually needed to lookup phases by areaid.
Also reverted change in SendQuestUpdate

roc13x committed:
Fix warning

Aokromes commented:
@Shauren this can be merged?

roc13x committed:
Fix possible crash

Faq commented:
/Me summons @shauren and @joschiwald

Lopfest commented:
this PR fixes #21196 and almost #18190
please can you verify and add following code in `Object.cpp` in `SetInPhase` so it fixes also #18190?
https://pastebin.com/mASa9Vvs
you can also fix log message here
`TC_LOG_ERROR("sql.sql", "%s Area %u does not have phase %u.", cond->ToString().c_str(), cond->SourceGroup, cond->SourceEntry);`
cond->SourceGroup and cond->SourceEntry are swapped

msoky commented:
any news?

tkrokli commented:
Good question, but I don't see any recent development here.

BTW, is this PR affected by **Core/Object: fix logic in WorldObject::UpdateAreaAndZonePhase() #21214** and its commit https://github.com/TrinityCore/TrinityCore/commit/5b90538919cdf8b1984049bd203104128bec5bf2 ?

Aokromes commented:
Conflicting files
src/server/game/Entities/Object/Object.cpp

tkrokli commented:
Hardly any activity from @roc13x since October 2017 (view the user page).

Shauren commented:
I started working on a bunch of phase related optimizations that will make this PR obsolete (you can see it here, accidentally pushed new files for this https://github.com/TrinityCore/TrinityCore/commit/bb718b557862057524778414ee728bf1923a14a4#diff-26a3ae191cac89dc294ec6b735af10ba)

roc13x commented:
Ok good to know. Saves me reviewing this tonight since 5b90538919cdf8b1984049bd203104128bec5bf2 happened
